src/main.cpp

```diff
@@ -403,39 +403,280 @@ bool FullSyncWithDPORNodes()
             return true;
 }
 
-double GetPoSKernelPS()
+double GetEstimatedNetworkWeight(unsigned int nPoSInterval)
+{
+    // The number of stakes to include in the average has been reduced to 40 (default) from 72. 72 stakes represented 1.8 hours at
+    // standard spacing. This is too long. 40 blocks is nominally 1 hour.
+    double result;
+
+    // The constant below comes from (MaxHash / StandardDifficultyTarget) * 16 sec / 90 sec. If you divide it by 80 to convert to GRC you
+    // get the familiar 9544517.40667
+    result = 763561392.533 * GetAverageDifficulty(nPoSInterval);
+    if (fDebug10) LogPrintf("Network Weight = %f", result);
+    if (fDebug10) LogPrintf("Network Weight in GRC = %f", result / 80.0);
+
+    return result;
+}
+
+double GetAverageDifficulty(unsigned int nPoSInterval)
 {
-    int nPoSInterval = 72;
-    double dStakeKernelsTriedAvg = 0;
-    int nStakesHandled = 0, nStakesTime = 0;
+    /*
+     * Diff is an inverse metric. It is essentially the reciprocal of the probability of staking
+     * (compared to standard target instead of maxhash). The diff is recorded in each block, and
+     * its inverse is a proxy for the coinweight selected by the staker. Therefore averaging the
+     * reciprocal of the diffs and then reinverting to determine the average diff is the correct way
+     * to do this calculation. This is called a harmonic mean in math.
+     *
+     * Also... The number of stakes to include in the average has been reduced to 40 (default) from 72.
+     * 72 stakes represented 1.8 hours at standard spacing. This is too long. 40 blocks is nominally 1 hour.
+     */
+     
+    double dDiff = 1.0;
+    double dReciprocalDiffSum = 0.0;
+    unsigned int nStakesHandled = 0;
+    double result;
 
-    CBlockIndex* pindex = pindexBest;;
-    CBlockIndex* pindexPrevStake = NULL;
+    CBlockIndex* pindex = pindexBest;
 
     while (pindex && nStakesHandled < nPoSInterval)
     {
         if (pindex->IsProofOfStake())
         {
-            dStakeKernelsTriedAvg += GetDifficulty(pindex) * 4294967296.0;
-            nStakesTime += pindexPrevStake ? (pindexPrevStake->nTime - pindex->nTime) : 0;
-            pindexPrevStake = pindex;
             nStakesHandled++;
+            dDiff = GetDifficulty(pindex);
```

Comment:
You can slim this down slightly to:

```
// dDiff should never be zero, but just in case, skip the block and move to the next one. 
dDiff = GetDifficulty(pindex);
if (!dDiff)
    continue;

dReciprocalDiffSum += 1 / dDiff;
nStakesHandled++;
if (fDebug10) LogPrintf("dDiff = %f", dDiff);
if (fDebug10) LogPrintf("nStakesHandled = %u", nStakesHandled);
```

Reply:
I agree.

Reply:
Ugh. Could create an infinite loop if someone specifies an interval of one, and the first block is bad.